File: backend/src/common/constants/feature-map.ts
export interface StudentFeatures {
  gender?: string;
  age?: number;
  logins?: number;
  totalHoursInModuleArea?: number;
  percentOfAverageHours?: number;
  presence?: number;
  absence?: number;
  percentAttended?: number;
  attendingFromHome?: string;
  distanceToUniversityKm?: number;
  polar4Quintile?: number;
  polar3Quintile?: number;
  adultHe2001Quintile?: number;
  adultHe2011Quintile?: number;
  tundraMsoaQuintile?: number;
  tundraLsoaQuintile?: number;
  gapsGcseQuintile?: number;
  gapsGcseEthnicityQuintile?: number;
  uniConnectTargetWard?: string;
}

export type FeatureType = 'numeric' | 'categorical';

export interface FeatureDefinition {
  key: keyof StudentFeatures;
  rawNames: string[];
  displayName: string;
  type: FeatureType;
}

// This mapping is the single source of truth for:
// raw CSV/JSON headers -> internal camelCase key -> UI label.
export const FEATURE_DEFINITIONS: FeatureDefinition[] = [
  { key: 'gender', rawNames: ['Gender'], displayName: 'Gender', type: 'categorical' },
  { key: 'age', rawNames: ['Age'], displayName: 'Age', type: 'numeric' },
  { key: 'logins', rawNames: ['Logins'], displayName: 'Logins', type: 'numeric' },
  {
    key: 'totalHoursInModuleArea',
    rawNames: ['Total Hours in Module Area'],
    displayName: 'Total Hours in Module Area',
    type: 'numeric',
  },
  {
    key: 'percentOfAverageHours',
    rawNames: ['% of Average Hours', 'Percent of Average Hours'],
    displayName: '% of Average Hours',
    type: 'numeric',
  },
  { key: 'presence', rawNames: ['Presence'], displayName: 'Presence', type: 'numeric' },
  { key: 'absence', rawNames: ['Absence'], displayName: 'Absence', type: 'numeric' },
  {
    key: 'percentAttended',
    rawNames: ['Percent Attended', '% Attended'],
    displayName: 'Percent Attended',
    type: 'numeric',
  },
  {
    key: 'attendingFromHome',
    rawNames: ['attending from home?', 'Attending from home?'],
    displayName: 'Attending from Home',
    type: 'categorical',
  },
  {
    key: 'distanceToUniversityKm',
    rawNames: ['distance to university (km)', 'Distance to university (km)'],
    displayName: 'Distance to University (km)',
    type: 'numeric',
  },
  {
    key: 'polar4Quintile',
    rawNames: ['POLAR4 Quintile'],
    displayName: 'POLAR4 Quintile',
    type: 'numeric',
  },
  {
    key: 'polar3Quintile',
    rawNames: ['POLAR3 Quintile'],
    displayName: 'POLAR3 Quintile',
    type: 'numeric',
  },
  {
    key: 'adultHe2001Quintile',
    rawNames: ['Adult HE 2001 Quintile'],
    displayName: 'Adult HE 2001 Quintile',
    type: 'numeric',
  },
  {
    key: 'adultHe2011Quintile',
    rawNames: ['Adult HE 2011 Quintile'],
    displayName: 'Adult HE 2011 Quintile',
    type: 'numeric',
  },
  {
    key: 'tundraMsoaQuintile',
    rawNames: ['TUNDRA MSOA Quintile'],
    displayName: 'TUNDRA MSOA Quintile',
    type: 'numeric',
  },
  {
    key: 'tundraLsoaQuintile',
    rawNames: ['TUNDRA LSOA Quintile'],
    displayName: 'TUNDRA LSOA Quintile',
    type: 'numeric',
  },
  {
    key: 'gapsGcseQuintile',
    rawNames: ['Gaps GCSE Quintile'],
    displayName: 'Gaps GCSE Quintile',
    type: 'numeric',
  },
  {
    key: 'gapsGcseEthnicityQuintile',
    rawNames: ['Gaps GCSE Ethnicity Quintile'],
    displayName: 'Gaps GCSE Ethnicity Quintile',
    type: 'numeric',
  },
  {
    key: 'uniConnectTargetWard',
    rawNames: ['Uni Connect target ward'],
    displayName: 'Uni Connect Target Ward',
    type: 'categorical',
  },
];

export const FEATURE_KEY_SET = new Set(FEATURE_DEFINITIONS.map((feature) => feature.key));

export const DISPLAY_NAME_BY_KEY: Record<string, string> = FEATURE_DEFINITIONS.reduce(
  (accumulator, feature) => {
    accumulator[feature.key] = feature.displayName;
    return accumulator;
  },
  {} as Record<string, string>,
);

File: ml-service/app/feature_map.py
from dataclasses import dataclass
from typing import Dict, List, Literal

FeatureType = Literal["numeric", "categorical"]


@dataclass(frozen=True)
class FeatureDefinition:
    key: str
    raw_names: List[str]
    display_name: str
    feature_type: FeatureType


FEATURE_DEFINITIONS: List[FeatureDefinition] = [
    FeatureDefinition("gender", ["Gender"], "Gender", "categorical"),
    FeatureDefinition("age", ["Age"], "Age", "numeric"),
    FeatureDefinition("logins", ["Logins"], "Logins", "numeric"),
    FeatureDefinition(
        "totalHoursInModuleArea",
        ["Total Hours in Module Area"],
        "Total Hours in Module Area",
        "numeric",
    ),
    FeatureDefinition(
        "percentOfAverageHours",
        ["% of Average Hours", "Percent of Average Hours"],
        "% of Average Hours",
        "numeric",
    ),
    FeatureDefinition("presence", ["Presence"], "Presence", "numeric"),
    FeatureDefinition("absence", ["Absence"], "Absence", "numeric"),
    FeatureDefinition(
        "percentAttended",
        ["Percent Attended", "% Attended"],
        "Percent Attended",
        "numeric",
    ),
    FeatureDefinition(
        "attendingFromHome",
        ["attending from home?", "Attending from home?"],
        "Attending from Home",
        "categorical",
    ),
    FeatureDefinition(
        "distanceToUniversityKm",
        ["distance to university (km)", "Distance to university (km)"],
        "Distance to University (km)",
        "numeric",
    ),
    FeatureDefinition("polar4Quintile", ["POLAR4 Quintile"], "POLAR4 Quintile", "numeric"),
    FeatureDefinition("polar3Quintile", ["POLAR3 Quintile"], "POLAR3 Quintile", "numeric"),
    FeatureDefinition(
        "adultHe2001Quintile",
        ["Adult HE 2001 Quintile"],
        "Adult HE 2001 Quintile",
        "numeric",
    ),
    FeatureDefinition(
        "adultHe2011Quintile",
        ["Adult HE 2011 Quintile"],
        "Adult HE 2011 Quintile",
        "numeric",
    ),
    FeatureDefinition("tundraMsoaQuintile", ["TUNDRA MSOA Quintile"], "TUNDRA MSOA Quintile", "numeric"),
    FeatureDefinition("tundraLsoaQuintile", ["TUNDRA LSOA Quintile"], "TUNDRA LSOA Quintile", "numeric"),
    FeatureDefinition("gapsGcseQuintile", ["Gaps GCSE Quintile"], "Gaps GCSE Quintile", "numeric"),
    FeatureDefinition(
        "gapsGcseEthnicityQuintile",
        ["Gaps GCSE Ethnicity Quintile"],
        "Gaps GCSE Ethnicity Quintile",
        "numeric",
    ),
    FeatureDefinition(
        "uniConnectTargetWard",
        ["Uni Connect target ward"],
        "Uni Connect Target Ward",
        "categorical",
    ),
]

DISPLAY_NAME_BY_KEY: Dict[str, str] = {feature.key: feature.display_name for feature in FEATURE_DEFINITIONS}
FEATURE_TYPES_BY_KEY: Dict[str, FeatureType] = {
    feature.key: feature.feature_type for feature in FEATURE_DEFINITIONS
}
NUMERIC_FEATURES: List[str] = [f.key for f in FEATURE_DEFINITIONS if f.feature_type == "numeric"]
CATEGORICAL_FEATURES: List[str] = [f.key for f in FEATURE_DEFINITIONS if f.feature_type == "categorical"]
FEATURE_KEYS: List[str] = [f.key for f in FEATURE_DEFINITIONS]


def normalize_feature_key(name: str) -> str:
    return " ".join(name.strip().lower().split())


RAW_OR_INTERNAL_TO_KEY: Dict[str, str] = {}
for definition in FEATURE_DEFINITIONS:
    RAW_OR_INTERNAL_TO_KEY[normalize_feature_key(definition.key)] = definition.key
    for raw_name in definition.raw_names:
        RAW_OR_INTERNAL_TO_KEY[normalize_feature_key(raw_name)] = definition.key
